firebase-perf/src/test/java/com/google/firebase/perf/application/FragmentStateMonitorTest.java

```diff
@@ -0,0 +1,86 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.perf.application;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+import android.app.Activity;
+import androidx.appcompat.app.AppCompatActivity;
+import androidx.core.app.FrameMetricsAggregator;
+import androidx.fragment.app.Fragment;
+import androidx.fragment.app.FragmentManager;
+import com.google.firebase.perf.FirebasePerformanceTestBase;
+import com.google.firebase.perf.config.ConfigResolver;
+import com.google.firebase.perf.config.DeviceCacheManager;
+import com.google.firebase.perf.transport.TransportManager;
+import com.google.firebase.perf.util.Clock;
+import com.google.firebase.perf.util.Timer;
+import com.google.firebase.perf.v1.TraceMetric;
+import com.google.testing.timing.FakeDirectExecutorService;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.stubbing.Answer;
+import org.robolectric.RobolectricTestRunner;
+
+/** Unit tests for {@link com.google.firebase.perf.application.FragmentStateMonitor}. */
+@RunWith(RobolectricTestRunner.class)
+public class FragmentStateMonitorTest extends FirebasePerformanceTestBase {
+
+  @Mock private Clock clock;
+  @Mock private Fragment mockFragment;
+  @Mock private FragmentManager mockfragmentManager;
+  @Mock private TransportManager mockTransportManager;
+  @Mock private AppCompatActivity mockActivity;
+  @Mock private AppStateMonitor appStateMonitor;
+  @Mock private FrameMetricsAggregator fma;
+
+  @Captor private ArgumentCaptor<TraceMetric> argTraceMetric;
+
+  private long currentTime = 0;
+
+  private Activity activity1;
+
+  @Before
+  public void setUp() {
+    currentTime = 0;
+    initMocks(this);
+    doAnswer((Answer<Timer>) invocationOnMock -> new Timer(currentTime)).when(clock).getTime();
+
+    DeviceCacheManager.clearInstance();
+    ConfigResolver.clearInstance();
+
+    ConfigResolver configResolver = ConfigResolver.getInstance();
+    configResolver.setDeviceCacheManager(new DeviceCacheManager(new FakeDirectExecutorService()));
+  }
+
+  @Test
```

Comment:
Can we add more test cases for this? This is an important part of validation and it would better to be exhaustive about the usecases we check here. Some scenarios that come to my mind:

1. Trace creation and stop for a particular fragment. What if we request for a trace again for the same fragment?
2. How do we verify if the hashmap for the fragment to trace is cleaned up as a part of the activity transition?
3. How do we handle scenarios where the length of the trace name for the fragment is way too high?
4. What happens if the fragment contains characters other than alphabets?

Think of more exhaustive usescases to make sure they are covered in unit tests as well.

Reply:
> 1. Trace creation and stop for a particular fragment. What if we request for a trace again for the same fragment?  

Could you elaborate on this?
> 2. How do we verify if the hashmap for the fragment to trace is cleaned up as a part of the activity transition?  

This would not be an unit test but an integration test, because it involves everything to be real (no Activity, Fragment mocks). 
> 3, 4 

These can be added as unit tests.

Reply:
Great idea on usecase 2 simulation, I feel like I could add a bunch of tests to previous fixes I've made using that idea.